### ISSUE TITLE
Must be reducible node exception

### ISSUE THREAD

**Issue by RiseeeAgainst** [dotnet/efcore#7935]
Must be reducible node exception

The reason of exception metod in class "NavigationRewritingExpressionVisitor"
method: "protected override Expression VisitBinary(BinaryExpression node)"

### Steps to reproduce

There is the link with zip archive of test project
https://drive.google.com/file/d/0B-WxIMR78dsVZ3p1TUJzS0lSM28/view?usp=sharing

I'm sorry for registering this issue in second time (#7832),  but now there is more useful information and no dependency with Odata Restier

 ctx.AppMetadata.Select((a) => new { oid = a.Oid, refobj = a.AdapterNavigation.OwnerAppName}).ToList(); - works fine without exception
ctx.AppMetadata.Select((a) => new { oid = a.Oid, refobj = a.AdapterNavigation.OwnerAppName, checkNull = a.AdapterNavigation == null }).ToList(); - throws an exception

### Further technical details
EF Core version: 1.1.0.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
IDE: Visual Studio 2015

**Comment by ajcvickers** [comment]
@maumar Assigning to you--could be a dupe of #7832

**Comment by maumar** [comment]
@RiseeeAgainst thank you for the improved repro - with the new information I was able to reproduce the issue, even on our current dev bits.

**Comment by maumar** [comment]
Simplified repro:

```cs
    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                //ctx.Database.EnsureDeleted();
                //ctx.Database.EnsureCreated();

                var r3 = ctx.AppMetadata.Select(a => new { refobj = a.AdapterNavigation.OwnerAppName, checkNull = a.AdapterNavigation == null }).ToList();
            }
        }
    }

    public class MyContext : DbContext
    {
        public DbSet<Adapter> Adapters { get; set; }
        public DbSet<AppMetadata> AppMetadata { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=.;Database=Repro7935;Trusted_Connection=True;MultipleActiveResultSets=true");
        }
    }

    public class Adapter
    {
        public Guid Id { get; set; }
        public string OwnerAppName { get; set; }
        public virtual ICollection<AppMetadata> AppMetadataAdapterNavigation { get; set; }
    }

    public class AppMetadata
    {
        public Guid Id { get; set; }
        public Guid? Adapter { get; set; }
        public virtual Adapter AdapterNavigation { get; set; }
    }
```

**Comment by smitpatel** [comment]
@maumar - Can you verify this? This seems like to be fixed in #7207

**Comment by maumar** [comment]
dupe of #7207, which is fixed in the current bits
